### ISSUE TITLE
Accessibility of icons

### ISSUE THREAD

**Issue by HugoGiraudel** [jacobwgillespie/styled-icons#27]
Hey there!

Alright, let’s talk about accessibility. Most of what I know comes from Florens Verschelde. He wrote a fantastic article about working with SVG icons and I highly suggest anyone to read it: https://fvsch.com/code/svg-icons/#section-html.

Here is a summary of the things that are necessary for accessible icons:

- The SVG element should have `focusable="false"` to avoid a double-focus issue on Internet Explorer and Edge.

- The SVG element should have `aria-hidden="true"` even when used as a standalone icon. The associated text should be passed alongside the icon, not within it.

- To provide context without displaying it visually, you can use a combination of CSS rules. Refer to https://hugogiraudel.com/2016/10/13/css-hide-and-seek/ for more information.

- In case the icon is used a standalone icon, it should have a `title` attribute to provide a tooltip on hover.

- It should be possible to pass the `role` and `aria-live` attributes.

Here is how we implemented it at N26 (I inlined and simplified the styles for the sake of the explanation):

```jsx
// 1. The default vertical-align is `baseline`, which leaves a few pixels of
//    space below the icon. Using `middle` prevents this.
// 2. Hide paths and strokes overflowing the viewbox.
const styles = {
  display: 'inline-block',
  height: '1em',
  width: '1em',
  color: 'inherit',
  fill: 'currentcolor',
  verticalAlign: 'middle', /* 1 */
  overflow: 'hidden' /* 2 */
}

// Hide content while keeping it accessible for users of assistive technologies.
// See: http://hugogiraudel.com/2016/10/13/css-hide-and-seek/
// See: https://github.com/twbs/bootstrap/pull/25197
const visuallyHidden = {
  border: '0 !important',
  clip: 'rect(1px, 1px, 1px, 1px) !important',
  height: '1px !important',
  overflow: 'hidden !important',
  paddingTop: '0 !important',
  paddingRight: '0 !important',
  paddingBottom: '0 !important',
  paddingLeft: '0 !important',
  position: 'absolute !important',
  whiteSpace: 'nowrap !important',
  width: '1px !important'
}

const Icon = props => {
  if (!props.icon) {
    return null
  }

  const SVG = (
    <svg
      styles={styles}
      viewBox='0 0 32 32'
      aria-hidden='true'
      focusable='false'
    >
      <use xlinkHref={`#icon-${props.icon}`} />
    </svg>
  )

  if (!props.label) {
    return SVG
  }

  // The `role` and `aria-live` attributes might be useful in case the icon
  // needs to provide further information to the accessibility tree such as
  // for a loading spinner.
  // See: https://codeburst.io/how-to-create-a-simple-css-loading-spinner-make-it-accessible-e5c83c2e464c
  return (
    <span
      title={props.label}
      role={props.role}
      aria-live={props['aria-live']}
    >
      {SVG}
      <span styles={visuallyHidden}>{props.label}</span>
    </span>
  )
}
```

And then we can use it like this:

```jsx
// Icon used alongside with text
<Icon icon='cogs' /> Settings
```

Or like this:

```jsx
// Standalone icon
<Icon icon='cogs' label='Settings' />
```

Or like this:

```jsx
// Standalone icon with extra contextual information
<Icon icon='loader' label='Loading…' role='alert' aria-live='assertive' />
```

I hope it helps. :)

**Comment by HugoGiraudel** [comment]
Thank you for fixing the possible focus bug on Internet Explorer and Edge, that’s already better!

Here is the thing about the `title` element: this is not the correct way to provide meaning to an icon. The `title` element is supposed to describe the graphics itself, hence why it is used for charts, diagrams or other imagery like this. For an icon, one does not want to describe the icon itself, but its meaning related to its usage. 

This is why the correct approach is to hide the icon from the accessibility tree, add visually hidden text next to it, as well as the `title` attribute for usability. I understand the hesitation on wrapping icons with a span, but unfortunately unless you do that, the accessibility layer will stay suboptimal at best.

**Comment by jacobwgillespie** [comment]
I'm going to close this issue for now, though I may revisit at some future point.  For now, this library provides raw access to the `<svg>` element, and there's a note in the README about wrapping.  But some future backwards-incompatible version might have built-in span-wrapping.
